Torch distributed support

amacati opened the pull request.

amacati committed:
initial commit

amacati committed:
maggy merge commit

moritzmeister commented:
closes #77, closes #78

RiccardoGrigoletto commented on maggy/core/executors/dist_executor.py:
```diff
@@ -0,0 +1,239 @@
+    """
+    client_addr = client.client_addr
+    host_port = (
+        client_addr[0] + ":" + "8081"
```
why is the port hardcoded?

amacati replied:
When I assigned ports outside a certain range during initial docker tests there were connection issues, so I just went with whatever worked. Gonna test if it works with all ports on the Spark cluster and change that line.

amacati committed:
heartbeat fix, random socket selection for pytorch master

amacati committed:
example notebook added

amacati commented:
Fixed the port assignment to be handled by the OS and the heartbeat issue. Also added a notebook in the examples folder so that you can easily check the code. Note that the code I inserted into the distributed driver is only present to make the heartbeat work, the mechanism will hopefully change during the next PR requests. Can you recheck the PR @RiccardoGrigoletto @moritzmeister ?

amacati committed:
fixed bugs, changed example to current version

moritzmeister commented on maggy/util.py:
```diff
@@ -234,3 +234,14 @@ def _validate_ml_id(app_id, run_id):
     return app_id, run_id
+
+
+def set_ml_id(app_id, run_id):
```
can you replace all occurrences of `experiment_utils._set_ml_id` with this new one here.

moritzmeister commented on maggy/core/executors/Executor.py:
```diff
@@ -0,0 +1,96 @@
+            exp_driver (Union[OptimizationDriver, AblationDriver, DistributedDriver]): Experiment
+            driver for the patching function.
+        """
+        assert type(exp_driver) in [
+            OptimizationDriver,
+            AblationDriver,
+            DistributedDriver,
+        ], f"Experiment driver type {type(exp_driver)} unsupported by Executor."
```
I tend to prefer `isinstance` instead of checking type equality, it's a bit safer since it checks also for inheritence. In this case it doesn't matter, just for the future 
```suggestion
        assert isinstance(exp_driver, [
            OptimizationDriver,
            AblationDriver,
            DistributedDriver,
        ]), f"Experiment driver type {type(exp_driver)} unsupported by Executor."
```

moritzmeister commented on maggy/distributed/patching.py:
```diff
@@ -0,0 +1,60 @@
+import torch
```
missing copyright

moritzmeister commented on maggy/core/experiment_driver/DistributedDriver.py:
```diff
@@ -0,0 +1,101 @@
+        self.result = {"best_val": "n.a.", "num_trials": 1, "early_stopped": 0}
+        self.job_start = None
+        self.executor_addresses = {}
+        self.spark_context = SparkSession.builder.getOrCreate().sparkContext
```
remove if not needed

moritzmeister commented on maggy/core/experiment_driver/DistributedDriver.py:
```diff
@@ -0,0 +1,101 @@
+import threading
+import queue
+
+from pyspark.sql import SparkSession
```
remove if not needed

moritzmeister reviewed (changes requested):
looks good, few remarks

amacati committed:
addressed comments
